CAMEL-10671: Adding Camel example project for the Ceylon JVM language

dimitrisli opened the pull request:
Adding Camel example for Ceylon 1.3.3. 

Followed the path of the Kotlin example as instructed on the ticket. 

Compile/Runs without errors and I'm able to access Jetty's localhost:8080 for the welcome message.

Things I couldn't get around: 

1.  The module.ceylon file cannot refer to Maven POM versions/properties unfortunately needing a hardcoded version of the used libs ([double-checked on the Ceylon Gitter channel too](https://gitter.im/ceylon/user?at=5adda26e62316e0505eceb43))
2. I had to hardcode the version of camel-http-common since the 2.22.0-SNAPSHOT does not exist

dmvolod commented on examples/camel-example-ceylon/pom.xml:
```diff
@@ -0,0 +1,114 @@
+    <dependency>
+      <groupId>org.apache.camel</groupId>
+      <artifactId>camel-http-common</artifactId>
+      <version>2.21.0</version>
```
This is not needed by default

dmvolod commented on examples/camel-example-ceylon/pom.xml:
```diff
@@ -0,0 +1,114 @@
+          <dependency>
+            <groupId>org.apache.camel</groupId>
+            <artifactId>camel-http-common</artifactId>
+            <version>2.21.0</version>
```
You can use ${project.version} here

dmvolod commented on examples/camel-example-ceylon/pom.xml:
```diff
@@ -0,0 +1,114 @@
+          <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>3.1.0</version>
```
You can use ${javax.servlet-api-version} here

dmvolod commented on examples/camel-example-ceylon/src/main/ceylon/org/apache/camel/example/module.ceylon:
```diff
@@ -0,0 +1,22 @@
+native("jvm")
+module org.apache.camel.example "1.0.0" {
+
+    import maven:"org.apache.camel:camel-core" "2.22.0-SNAPSHOT";
```
If you don't want to hardcode camel components version inside .ceylone module, it's possible to use some template file and fill it on the build stage inside maven.

dimitrisli replied:
I tried a few things but it didn't work. The closest one I got was the Filtering capability of the Maven Resources plugin.(https://maven.apache.org/plugins/maven-resources-plugin/examples/filter.html). It works well in what we want it to do: pointing at the package where the module.ceylon is included it is indeed changing the ${project.version} and store it under target/classes. Problem is that the Ceylon compiler (under the plugin) is not considering the target entry but rather the one under sources. I tried with the resources config of the Ceylon plugin too to no avail

dimitrisli commented:
Replaced version 3.1.0 with existing property javax.servlet-api-version

dimitrisli commented:
Indeed - my 2nd point on the description about camel-http-common is not valid since after mvn install it's there locally.. thanks for suggesting. I've fixed the versions as needed (in the dependencies section it gets inherited and under the plugin i specified ${project.version}

dmvolod commented on examples/camel-example-ceylon/src/main/ceylon/org/apache/camel/example/run.ceylon:
```diff
@@ -0,0 +1,32 @@
+
+    CamelMain main = CamelMain();
+    main.addRouteBuilder(MyRouteBuilder());
+    main.run(createJavaStringArray(process.arguments));
```
it's not about just parameters, it's necessary to call main.main(...) method or add System.exit in ceylon code.
Please look at this code as basis https://github.com/apache/camel/blob/241f7bd00ba3f3a2a556ec74de15e73901051ba6/camel-core/src/main/java/org/apache/camel/main/Main.java#L48

dimitrisli replied:
Thanks that indeed fixed the error on ctrl+c!

dimitrisli commented:
Your idea about the template maven file move during build helped me take my previous attempt a step further. 

Basically: 

1) I removed all together the **module.ceylon** file from the **src/** directory and into the resources replacing the version with the **${project.version}** placeholders 

2) as before **resources:resources** is doing its job replacing the placeholders and placing the file under **resources/**

3) now I've added the **maven ant plugin** early in the maven cycle to copy that file from the **resources/** back to the **src/** directory

dimitrisli commented:
So now I think it looks good. Let me know if you spot anything else needed to address.

dmvolod commented on examples/camel-example-ceylon/pom.xml:
```diff
@@ -0,0 +1,152 @@
+
+      <plugin>
+        <artifactId>maven-antrun-plugin</artifactId>
+        <version>1.4</version>
```
maven-antrun-plugin-version already defined in parent pom, you don't need to define it here

dmvolod commented on examples/camel-example-ceylon/ReadMe.md:
```diff
@@ -0,0 +1,29 @@
+
+To build this project use
+
+    mvn clean resources:resources install
```
The resources phase is not needed for build.

dimitrisli commented:
Addressed all 3 points raised in individual commits:

1. Removed Maven Ant plugin version since inherited from parent pom
2. In the instructions moved the 'resources:resources' from building to running
3. Added .gitignore with module.ceylon in it to cater for the copying on the src/ dir that is under git

dmvolod commented:
@dimitrisli could you please to squash commits
@davsclaus could you please to look at this PR and merge it if possible
Thank you.

dimitrisli commented:
@dmvolod not sure if I can squash pushed commits without --force?

dmvolod commented:
@dimitrisli, it's not a problem to update PR with -force option. This works well on GitHub.

davsclaus commented:
Hmm I have not used Celyon before and it seems a bit odd you need to have source code files with dependency settings and maven pom.xml as well and then need all kind of maven ant trick to update version numbers in the source code file.

Can't the celyon maven plugin do more to help there, and is that celyon module file needed?

dimitrisli commented:
I've done one final polish commit where I've removed unnecessary dependencies and ensured there is nowhere hardcoded versions. Will look into squashing all these commits into a single one.

@davsclaus Definitely not a Ceylon expert here as still learning. But my understanding is: Ceylon [offers an out-of-the-box module system support built-in the language](https://ceylon-lang.org/documentation/1.1/reference/structure/module/) as opposed to Java (pre project Jigsaw) / Kotlin / Scala. This first-class module system support makes Ceylon module descriptor file(s) obligatory as they come first and in general there is no need to rely on say Maven/Gradle at all. In fact Maven support was greatly improved in version 1.3.2 via the Ceylon Maven plugin to facilitate the up-until-then somehow clunky Ceylon/Java interop via Maven.

Now on the question 'can we avoid hardcoding the versions in the module descriptor file' that would saves us from the replace-copy-paste hooks we are doing here via Maven resources+ant plugins: [It doesn't seem to be the case based on my Gitter channel question](https://gitter.im/ceylon/user?at=5adda26e62316e0505eceb43). The best we can do is [group the same version refs within the module descriptor](https://gitter.im/ceylon/user?at=5adf0ece2b9dfdbc3ac2bc1c). Also there is [this somehow 'heated' discussion on why there is the need to hardcode Maven version dependency numbers in the first place within the module descriptor](https://github.com/eclipse/ceylon/issues/7030).

WillemJiang commented:
Thanks for this Ceylon example,  maybe we can work on some sophisticated DSL with Ceylon later.

PascalSchumacher commented:
For some reason this does not compile on builds.apache.org:
 
```
[INFO] -----------< org.apache.camel.example:camel-example-ceylon >------------
[INFO] Building Camel :: Example :: Ceylon 2.22.0-SNAPSHOT            [655/768]
...
[ERROR] Compilation error at (21,5) in /home/jenkins/jenkins-slave/workspace/Camel_master-CIGY56HHEDNGVJBWJRRHBZ64GKNJTCSYDKOPJS46W5IZTMAS3I2Q/examples/camel-example-ceylon/src/main/ceylon/org/apache/camel/example/module.ceylon:cannot find module artifact 'maven:org.apache.camel:camel-core-2.22.0-SNAPSHOT(.car|.jar)'
  	- dependency tree: 'org.apache.camel.example/2.22.0-SNAPSHOT' -> 'org.apache.camel:camel-core/2.22.0-SNAPSHOT'
[ERROR] Compilation error at (22,5) in /home/jenkins/jenkins-slave/workspace/Camel_master-CIGY56HHEDNGVJBWJRRHBZ64GKNJTCSYDKOPJS46W5IZTMAS3I2Q/examples/camel-example-ceylon/src/main/ceylon/org/apache/camel/example/module.ceylon:cannot find module artifact 'maven:org.apache.camel:camel-jetty-2.22.0-SNAPSHOT(.car|.jar)'
  	- dependency tree: 'org.apache.camel.example/2.22.0-SNAPSHOT' -> 'org.apache.camel:camel-jetty/2.22.0-SNAPSHOT'
```

For details see: https://builds.apache.org/view/C/view/Apache%20Camel/job/Camel/job/master/264/consoleFull

PascalSchumacher commented:
I know nothing about Ceylon, but I guess the problem is that camel snapshots are in a custom maven repository:  https://repository.apache.org/content/repositories/snapshots/

Seems like the only way to make Ceylon itself aware of this is adding a custom `settings.xml` (see: https://ceylon-lang.org/documentation/1.3/reference/repository/maven/).

I think a better way to solve may be to import the maven dependencies as Ceylon modules with the `ceylon-maven-plugin` (see: https://github.com/ceylon/ceylon-maven-plugin#import-maven-dependencies-as-ceylon-modules).

Maybe this can also be used to remove the `module.ceylon` manipulation with the `maven-ant-run-plugin`?

dimitrisli commented:
sorry I had a brief look last night and will have more time this afternoon and w/e. In the meantime if this is holding off the whole build please revert the commit.

WillemJiang commented:
I just found import maven dependencies as ceylon modules is a solution.  I will commit a quick fix for it in a min.

PascalSchumacher commented:
Finally got it to compile on builds.apache.org with these additional changes: https://github.com/apache/camel/commit/943f6fd83d6917a1ec70ab1cfd41d0a8d62a73db and https://github.com/apache/camel/commit/ed69c80eab1f33ee6606078e74d4cb1371453bb6

dimitrisli commented:
many thanks for your help guys.

PascalSchumacher commented:
No the build fails later on:

```
[ERROR] Failed to execute goal org.ceylon-lang:ceylon-maven-plugin:1.3.3:import-dependency (dependency) on project camel-example-ceylon: Cannot import module maven:org.apache.camel.camel-core/2.22.0-SNAPSHOT: Jar file /home/jenkins/jenkins-slave/workspace/Camel_master-CIGY56HHEDNGVJBWJRRHBZ64GKNJTCSYDKOPJS46W5IZTMAS3I2Q/camel-core/target/classes is a directory (must be a file) -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.ceylon-lang:ceylon-maven-plugin:1.3.3:import-dependency (dependency) on project camel-example-ceylon: Cannot import module maven:org.apache.camel.camel-core/2.22.0-SNAPSHOT
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:213)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:50)
Caused by: org.apache.maven.plugin.MojoExecutionException: Cannot import module maven:org.apache.camel.camel-core/2.22.0-SNAPSHOT
    at com.redhat.ceylon.maven.CeylonImportDependencyMojo.execute (CeylonImportDependencyMojo.java:127)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:50)
Caused by: com.redhat.ceylon.tools.importjar.ImportJarException: Jar file /home/jenkins/jenkins-slave/workspace/Camel_master-CIGY56HHEDNGVJBWJRRHBZ64GKNJTCSYDKOPJS46W5IZTMAS3I2Q/camel-core/target/classes is a directory (must be a file)
    at com.redhat.ceylon.tools.importjar.CeylonImportJarTool.checkReadableFile (CeylonImportJarTool.java:254)
    at com.redhat.ceylon.tools.importjar.CeylonImportJarTool.assertJar (CeylonImportJarTool.java:236)
    at com.redhat.ceylon.tools.importjar.CeylonImportJarTool.initialize (CeylonImportJarTool.java:187)
    at com.redhat.ceylon.maven.CeylonImportDependencyMojo.execute (CeylonImportDependencyMojo.java:124)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:208)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:154)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:146)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:290)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:194)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.apache.maven.wrapper.BootstrapMainStarter.start (BootstrapMainStarter.java:39)
    at org.apache.maven.wrapper.WrapperExecutor.execute (WrapperExecutor.java:122)
    at org.apache.maven.wrapper.MavenWrapperMain.main (MavenWrapperMain.java:50)
```

dimitrisli commented:
After a few attempts I cannot reproduce locally to further troubleshoot. I cannot see the maven targets Jenkins is running but based on the surroundings I'm guessing it's doing: 1) mvn clean install , 2) chestyle, 3) mvn package surfire:test. The (1) and (2) have no issues now and the error in on the (3). I've noticed that on (3) we don't run "mvn clean" which we do on (1). Can we add that before package surfire:test? Or since there are no tests on this example project skip it altogether?

dimitrisli commented:
Like for instance in this example project [there is this token defined in the child pom props](https://github.com/apache/camel/blob/3e07a6385b8ba57ba4eb666bd7d2b13ea53306fb/examples/camel-example-spring-boot-geocoder/pom.xml#L37) that is preventing the tests to execute so we can try this.

dimitrisli commented:
Submitted new PR for this: https://github.com/apache/camel/pull/2319

WillemJiang commented:
@dimitrisli & @PascalSchumacher  I just have question about difference between  
`maven:org.apache.camel.camel-core/2.22.0-SNAPSHOT` and `org.apache.camel.camel-core/2.22.0-SNAPSHOT`. 
If we add the camel related module into ceylon managed repo,  it doesn't make sense we got  this kind of error.

BTW, I didn't see the log related jars in the module.ceylon, is it OK in that way ?

PascalSchumacher commented:
>I cannot see the maven targets Jenkins is running

The targets run by Jenkins are defined here: https://github.com/apache/camel/blob/master/Jenkinsfile

@WillemJiang:
My understanding is that with `shared import maven:"org.apache.camel:camel-core" "${project.version}";` Ceylon uses maven-central and what repositories defined in `settings.xml` to resolve the dependency.
With `shared import "org.apache.camel.camel-core" "${project.version}";`  Ceylon resolves the dependency using the Ceylon repository or the local module directory (Should be `./target/modules` when using the ceylon-maven-plugin. The `import-dependency` goal of the ceylon-maven-plugin writes the imported maven dependencies to this directory.)

dimitrisli commented:
That's my understanding as well. The "maven:" prefix in the ceylon.module is going down the maven repo route whereas the standard non-maven is looking at the target/modules directory for .car files (=jar + module descriptor - a custom Ceylon module archive). Now with the ceylon:import-dependency plugin and the force option there are jar files within the modules directory. 

Now the error is coming from the [Ceylon import-jar](https://ceylon-lang.org/documentation/1.0/reference/tool/ceylon/subcommands/ceylon-import-jar.html) functionality that is trying to load the jar thinking that it is under target/classes rather looking within the modules directory. As @WillemJiang I also don't understand why in the error we can see the "maven:" prefix though we don't explicitly define it as such.

About the logging jars if there was an issue we would know at runtime but there wasn't any such issue.

Why don't we try the skipTests parameter? If I understood it correctly judging by the other example components that are using it, it is skipping the [stage('Test')](https://github.com/apache/camel/blob/master/Jenkinsfile#L76) which is where we are getting this error although we don't have a test.

dimitrisli commented:
@PascalSchumacher thanks - just raised : https://github.com/apache/camel/pull/2325 (under wrong ticket missed the last digit but edited the PR subject)
